### ISSUE TITLE
torch.special tracking issue

### ISSUE THREAD

**Issue by mruberry** [pytorch/pytorch#50345]
This issue is a roll-up of torch.special discussion, requests and issues. In particular, it is intended to track porting existing PyTorch functions analogous to scipy.special functions to a new torch.special namespace, and scipy.special function requests.

**Porting to torch.special**

- [ ] Create a torch.special namespace
- [ ] Create a doc for the torch.special module
- [ ] Identify all scipy.special functions in PyTorch
- [ ] Add i0
- [ ] Add xlogy
- [ ] Add logsumexp
- [ ] Add sinc
- [ ] Add round
- [ ] Add expm1
- [ ] Add log1p
- [ ] Add radian
- [ ] Add exp2
- [ ] Add exp10
- [ ] Add log_softmax
- [ ] Add softmax
- [ ] Add erf
- [ ] Add erfc
- [ ] Add erfinv
- [ ] Add digamma
- [ ] Add polygamma

**Requests for scipy.special functions**

**Comment by vadimkantorov** [comment]
Many of these funcions like softmax / log1p / round may be of more common use in PyTorch than they were originally in SciPy. Will this mean deprecating them in torch namespace and having to add everywhere import torch.special ?

We already have F.softmax, torch.softmax - a lot of confusion about torch.* versions and F.* verisons...

**Comment by mruberry** [comment]
Good question, @vadimkantorov. My thinking was to support aliases. We don't have to document the aliases so there can be a canonical torch.foo or torch.special.foo, however.

**Comment by SsnL** [comment]
What about `special.iv` and `special.ive`?

**Comment by kshitij12345** [comment]
> What about special.iv and special.ive?

Will consider this as request for the same. 
Have updated the `Requests for scipy.special functions`.

**Comment by nicola-decao** [comment]
Hi @vadimkantorov @kshitij12345 @mruberry I was wondering if we get in contact with someone from `scipy`. I saw that the `betainc` is implemented here and is quite easy to port to torch I think https://github.com/scipy/scipy/blob/master/scipy/special/cephes/incbet.c . I may have time to do it so I was wondering what would happen if port that code for `torch`

**Comment by kshitij12345** [comment]
@nicola-decao that'd great!! Feel free to tag me for review. Also feel free ping me (on the PR) if there is any confusion regarding the same.

Thanks!!

**Comment by meaneych** [comment]
Hello. Could we also add special.kv. https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.kv.html
The modified Bessel second kind appears in some statistical distributions. https://en.wikipedia.org/wiki/Generalized_inverse_Gaussian_distribution

Noob pytorch question? How are these scipy.special functions being added to pytorch? 
https://pytorch.org/tutorials/advanced/numpy_extensions_tutorial.html
Would a backward method to Function be needed for kv (both with respect to order nu and argument z)?
https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/20/ShowAll.html

I could have a crack at implementing, but might need more direction.

**Comment by mruberry** [comment]
> Hello. Could we also add special.kv. https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.kv.html
> The modified Bessel second kind appears in some statistical distributions. https://en.wikipedia.org/wiki/Generalized_inverse_Gaussian_distribution

We should definitely add it to the list of requests above, thank you for the suggestion.

> Noob pytorch question? How are these scipy.special functions being added to pytorch?
> https://pytorch.org/tutorials/advanced/numpy_extensions_tutorial.html

The best way to learn how to add an op is to look at the PRs implementing similar operators. Many of them are linked above.

> Would a backward method to Function be needed for kv (both with respect to order nu and argument z)?
> https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/20/ShowAll.html

Implementing support for autograd isn't always required, but if the function is differentiable that'd be great!

**Comment by meaneych** [comment]
> We should definitely add it to the list of requests above, thank you for the suggestion. Can you elaborate specifically on what you would use a special.kv implementation for, @meaneych?
>
> > @mruberry Thanks for feedback/support. In terms of use case for torch.special.kv()...I am trying to implement a number of count regression models in PyTorch. I created a small GitHub repository which describes the project objectives. See link: https://github.com/meaneych/PyTorch_CountDistribution_Examples 
Basically, the second order modified Bessel function (scipy.special.kv()) function appears in a number of these log density functions, which specify the loss/objective for these count regression models. For example, see PIG density of SICHEL density functions defined on pages 196-198 of the following manual (I am trying to port these codes to PyTorch):
http://www.gamlss.com/wp-content/uploads/2013/01/gamlss-manual.pdf 
>
> Noob pytorch question? How are these scipy.special functions being added to pytorch?
>
> > The best way to learn how to add an op is to look at the PRs implementing similar operators. Many of them are linked above.
> 
> > > Sorry @mruberry could you please provide a URL (link to file) containing a complete example of porting a scipy.special function to torch.special (sorry am very new to software eng, GitHub, etc.)?? I will try to follow the template provided exactly. Currently I have downloaded both Scipy and PyTorch repositories from GitGub. I have used `grep` to try to understand where `scipy.kv()` function is being used/called in scipy.special. From scipy side of things I have found a number of source code files (in C and Fortran) in the following locations:  
/scipy/special/cephes/scipy_iv.c  
/scipy/special/amos/zbesk.f  
/scipy/special/amos/zbknu.f  
etc.  
From the PyTorch side of things I tried to understand where your other Bessel functions implemented to date are being housed/called? I have found the following file (which calls torch.special.i0): /pytorch/aten/src/ATen/native/UnaryOps.cpp
>>> 
```
// special_i0, alias for i0
Tensor& special_i0_out(const Tensor& self, Tensor& result) { return at::i0_out(result, self); }
Tensor special_i0(const Tensor& self) { return self.i0(); }
```
>>> Am not sure if this is what is needed to port scipy.special to torch.special???

> > Would a backward method to Function be needed for kv (both with respect to order nu and argument z)?
> > https://functions.wolfram.com/Bessel-TypeFunctions/BesselK/20/ShowAll.html
> 
> Implementing support for autograd isn't always required, but if the function is differentiable that'd be great!
>
>>> Above WolfRam link gives derivative of BesselK function in terms of argument (z). The derivative is related to other BesselK functions. The derivative of the BesselK function in terms of order (nu) is given in link below (again it is a function of other Bessel functions). I suppose although analytic derivatives exists; the functions describing these derivatives are not available in PyTorch (because they are defined in terms of other Bessels)?? What do you think??  
https://dlmf.nist.gov/10.38  
>>> I think scipy.special also has functions defining Bessel derivatives? See link/URL below (zeroes of Bessel function).  
https://docs.scipy.org/doc/scipy/reference/special.html

**Comment by mruberry** [comment]
Take a look at https://github.com/pytorch/pytorch/pull/43132, @meaneych, although porting a special function might be a tricky first PR and it might be easier to start elsewhere.

There's an additional wrinkle, too. These special functions typically increase PyTorch's build time and CUDA context size. We're looking at some approaches to mitigate those issues, but until they're resolved we should only add CPU implementations of these operations and only for commonly used dtypes like float and double (or complex float and complex double if appropriate).

**Comment by slihn** [comment]
There are two more functions that are in scipy but not in torch: gammaincinv and gammainccinv.
They are the inverse functions of gammainc and gammaincc.
Please consider them for addition once you resolve the cuda context size issue.

https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.gammainccinv.html#r995da0bf15ea-dlmf

**Comment by ifsheldon** [comment]
Can you please consider adding bessel functions `j0`, `j1` and `jv`? We're doing microscopy simulation with PyTorch, but now it has only `i0` and other variants of modified bessel function. We're now resorting to naive adaption with PyTorch operators from Scipy.

**Comment by khushi-411** [comment]
Hi,
I'm planning to pick up `logerfc` and `logerfcx` for the next. I want to work on special functions this time.🤪 

May I go for it? Please let me know if anyone else is working on the same. I'd pick another one :) Also, I would love to get your guidance. (currently, I'm walking through previously mentioned PR's and Wiki).

Thanks!

cc: @mruberry @kshitij12345

**Comment by Dbhasin1** [comment]
Hello, is it alright if I pick up `gammaincinv` and `gammainccinv` if no one's already working on them? Thanks!
cc: @mruberry @kshitij12345

**Comment by mruberry** [comment]
> Hello, is it alright if I pick up `gammaincinv` and `gammainccinv` if no one's already working on them? Thanks! cc: @mruberry @kshitij12345

There are no "reservations" for work and anyone is free to work on anything at any time.
